osu.Game/Online/OnlineViewLayer.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Online.API;
+using osu.Game.Online.Placeholders;
+using osuTK.Graphics;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A layer which displays on top of content which require a local user to be logged in.
+    /// Blocks input to the underlying content, dims the optional target content and displays a login placeholder when the user isn't logged in.
+    /// </summary>
+    public class OnlineViewLayer : VisibilityContainer, IOnlineComponent
+    {
+        protected LoadingSpinner LoadingSpinner;
+
+        private readonly Drawable viewTarget;
+
+        private readonly string placeholderMessage;
+
+        private Placeholder placeholder;
+
+        private const double transform_duration = 500;
+
+        [Resolved]
+        protected IAPIProvider API { get; private set; }
+
+        public OnlineViewLayer(string placeholderMessage, Drawable viewTarget)
+        {
+            this.placeholderMessage = placeholderMessage;
+            this.viewTarget = viewTarget;
+
+            RelativeSizeAxes = Axes.Both;
+        }
+
+        public override bool HandleNonPositionalInput => false;
+
+        protected override bool Handle(UIEvent e) => true;
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            Children = new Drawable[]
+            {
+                placeholder = new LoginPlaceholder(placeholderMessage)
```

Comment:
Doesn't really make sense to have a `LoginPlaceholder` be used here, in a general `OnlineViewLayer` - should be either abstracted away to just a `Placeholder` or renamed to reflect the new more-general purpose.